cmd/sgx_epchook/main.go

```diff
@@ -30,45 +32,82 @@ import (
 )
 
 const (
-	namespace  = "sgx.intel.com"
-	epc        = "epc"
-	pathPrefix = "/status/capacity"
+	namespace = "sgx.intel.com"
+	epc       = "epc"
 )
 
-type patchExtendedResource struct {
-	Op    string `json:"op"`
-	Path  string `json:"path"`
-	Value uint64 `json:"value"`
+type patchNodeOp struct {
+	Op    string      `json:"op"`
+	Path  string      `json:"path"`
+	Value interface{} `json:"value"`
 }
 
 func main() {
-	var register, affirm bool
+	var register, affirm, label, daemon bool
 	flag.BoolVar(&register, "register", false, "register EPC as extended resource")
 	flag.BoolVar(&affirm, "affirm", false, "return error if EPC is not available")
+	flag.BoolVar(&label, "node-label", false, "create node label")
+	flag.BoolVar(&daemon, "daemon", false, "run as a daemon")
 	flag.Parse()
 
+	klog.Infof("starting sgx_epchook")
+
 	// get the EPC size
 	var epcSize uint64
 	if cpuid.CPU.SGX.Available {
 		for _, s := range cpuid.CPU.SGX.EPCSections {
 			epcSize += s.EPCSize
 		}
 	}
+	klog.Infof("epc capacity: %d bytes", epcSize)
 
 	if epcSize == 0 && affirm {
 		klog.Fatal("SGX EPC is not available")
 	}
 
-	if register {
-		if err := registerExtendedResource(epcSize); err != nil {
-			klog.Fatal(err.Error())
-		}
-	} else {
+	if err := updateNode(epcSize, register, label); err != nil {
+		klog.Fatal(err.Error())
+	}
+
+	if !register {
 		fmt.Printf("%s/%s=%d", namespace, epc, epcSize)
 	}
+
+	if daemon {
+		klog.Info("waiting for termination signal")
+		term := make(chan os.Signal, 1)
+		signal.Notify(term, os.Interrupt, syscall.SIGTERM)
+		<-term
+	}
 }
 
-func registerExtendedResource(epcSize uint64) error {
+func updateNode(epcSize uint64, register, label bool) error {
+	// create patch payload
+	resource := fmt.Sprintf("%s~1%s", namespace, epc)
+	payload := []patchNodeOp{}
+	if register {
+		payload = append(payload, patchNodeOp{
+			Op:    "add",
+			Path:  "/status/capacity/" + resource,
+			Value: epcSize,
+		})
+	}
+	if label && epcSize > 0 {
+		payload = append(payload, patchNodeOp{
+			Op:    "add",
+			Path:  "/metadata/labels/" + resource,
+			Value: "enabled",
```

Comment:
making the label and the resource the same might be confusing. how about `sgx.intel.com/capable: true` (true is at least better than enabled).